File: Exercism Codes/Java/Satellite/Satellite.java
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Satellite {

    private int preorderIndex = 0;
    public Tree treeFromTraversals(List<Character> preorderInput, List<Character> inorderInput) {
        Set<Character> inOrder = new HashSet<>(inorderInput);
        Set<Character> preOrder = new HashSet<>(preorderInput);
        if(inorderInput.size() != inOrder.size() || preorderInput.size() != preOrder.size())
            throw new IllegalArgumentException("traversals must contain unique items");
        if(preorderInput.size() != inorderInput.size())
            throw new IllegalArgumentException("traversals must have the same length");
        if(!inOrder.equals(preOrder))
            throw new IllegalArgumentException("traversals must have the same elements");
        return new Tree(buildTree(inorderInput, preorderInput, 0, inorderInput.size() - 1));
    }

    Node buildTree(List<Character> inorderInput, List<Character> preorderInput, int start, int end) {
        if (start > end) return null;
        Node n = new Node(preorderInput.get(preorderIndex++));
        if (start == end) return n;
        int inorderIndex = search(inorderInput, start, end, n.value);
        n.left = buildTree(inorderInput, preorderInput, start, inorderIndex - 1);
        n.right = buildTree(inorderInput, preorderInput, inorderIndex + 1, end);
        return n;
    }

    int search(List<Character> inorderInput, int start, int end, char value) {
        int i;
        for (i = start; i <= end; i++) if (inorderInput.get(i) == value) return i;
        return i;
    }
}

File: Exercism Codes/Python/satellite.py
preorder_index = 0

def tree_from_traversals(preorder, inorder):
    inorder_input = set(inorder)
    preorder_input = set(preorder)
    if len(inorder_input) != len(inorder) or len(preorder_input) != len(preorder):
        raise ValueError("traversals must contain unique items")
    if len(preorder) != len(inorder):
        raise ValueError("traversals must have the same length")
    if inorder_input != preorder_input:
        raise ValueError("traversals must have the same elements")

    preorder_index = 0

    def build_tree(pre_order, in_order, start, end):
        nonlocal preorder_index
        if start > end:
            return {}
        n = {"v": pre_order[preorder_index], "l": {}, "r": {}}
        preorder_index += 1
        if start == end:
            return n
        inorder_index = search(in_order, start, end, n["v"])
        n["l"] = build_tree(pre_order, in_order, start, inorder_index - 1)
        n["r"] = build_tree(pre_order, in_order, inorder_index + 1, end)
        return n


    return build_tree(preorder, inorder, 0, len(inorder) - 1)

def search(inorder, start, end, n):
    i = start
    for i in range(start, end + 1):
        if inorder[i] == n:
            return i
    return i